File: joss/comparisons/cat-6-contextual-runtime/31-nullable-vs-optional/example-nullable-vs-optional.R
# Error Class 31: Nullable vs Optional Distinction
#
# R has no distinction between null and undefined. Both empty cells
# and missing columns become NA. No semantic difference.

library(tidyverse)

df <- tibble(
  patient_id = c("P001", "P002"),
  lab_value = c(100, NA)  # explicitly missing
)

# No way to distinguish "explicitly null" from "field doesn't exist"
print(is.na(df$lab_value[2]))  # TRUE
# Both → NA. No semantic distinction.

File: joss/comparisons/cat-6-contextual-runtime/31-nullable-vs-optional/example-nullable-vs-optional.py
"""
Error Class 31: Nullable vs Optional Distinction

Python has no distinction between null and undefined. Both empty cells
and missing columns become NaN. You can't tell whether a value was
explicitly missing or the field didn't exist.
"""
import pandas as pd
import numpy as np

df = pd.DataFrame({
    "patient_id": ["P001", "P002"],
    "lab_value": [100, np.nan],  # explicitly missing
})

# No way to distinguish "explicitly null" from "field doesn't exist"
print(pd.isna(df["lab_value"].iloc[1]))  # True
# Both empty cell and missing column → NaN. No semantic distinction.
